File: algo/109. Convert Sorted List to Binary Search Tree.py
# The space usage is not optimal
# We can use O(logn) extra space instead of O(n)
# Time complexity is still O(nlogn)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        temp = head
        nums = []
        while (temp):
            nums.append(temp.val)
            temp = temp.next
        return self.sortedArrayToBST(nums)

    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if len(nums) == 0:
            return
        if len(nums) == 1:
            return TreeNode(nums[0])
        if len(nums) == 2:
            node1 = TreeNode(nums[0])
            node2 = TreeNode(nums[1])
            node2.left = node1
            return node2
        if len(nums) == 3:
            node1 = TreeNode(nums[0])
            node2 = TreeNode(nums[1])
            node3 = TreeNode(nums[2])
            node2.left = node1
            node2.right = node3
            return node2
        midIndex = len(nums) // 2
        mid = TreeNode(nums[midIndex])
        left = self.sortedArrayToBST(nums[:midIndex])
        right = self.sortedArrayToBST(nums[midIndex+1:])
        mid.left = left
        mid.right = right
        return mid        

File: round3/109. Convert Sorted List to Binary Search Tree.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode SortedListToBST(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return new TreeNode(head.val);
        
        var pre = new ListNode();
        pre.next = head;
        var slow = head;
        var fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            pre = pre.next;
        }
        
        pre.next = null;
        TreeNode root = new TreeNode(slow.val);
        root.left = SortedListToBST(head);
        root.right = SortedListToBST(slow.next);
        return root;
    }
}